Add 'node' CLI command to list nodes and show their status.

cuongdo opened the pull request:
The `node` command has two subcommands, `ls` and `status`.

`cockroach node ls` provides the NodeID's for all nodes in the cluster. For example:

```
$ cockroach node ls
+----+
| id |
+----+
|  1 |
|  2 |
|  3 |
+----+
```

`node status` takes 0 or 1 arguments. With no arguments, `node status` shows the status for all nodes. For example:

```
$ cockroach node status
+----+-----------------+---------------------+---------------------+------------+-----------+-------------+--------------+--------------+---------------+-------------+--------------+
| id |     address     |     updated_at      |     started_at      | live_bytes | key_bytes | value_bytes | intent_bytes | system_bytes | leader_ranges | repl_ranges | avail_ranges |
+----+-----------------+---------------------+---------------------+------------+-----------+-------------+--------------+--------------+---------------+-------------+--------------+
|  1 | 127.0.0.1:26001 | 2016-01-13 22:05:42 | 2016-01-13 13:06:02 |   12847326 |    144151 |    14393513 |          329 |         2444 |             1 |           1 |            1 |
|  2 | 127.0.0.2:26002 | 2016-01-13 22:05:45 | 2016-01-13 13:06:05 |   12849693 |    144199 |    14396408 |           75 |         6829 |             4 |           4 |            4 |
|  3 | 127.0.0.3:26003 | 2016-01-13 22:05:44 | 2016-01-13 13:07:24 |   12847569 |    144175 |    14394021 |           75 |         6829 |             0 |           0 |            0 |
+----+-----------------+---------------------+---------------------+------------+-----------+-------------+--------------+--------------+---------------+-------------+--------------+
```

`node status <node ID>` shows the status for a single node. For example:

```
$ cockroach node status 1
+----+-----------------+---------------------+---------------------+------------+-----------+-------------+--------------+--------------+---------------+-------------+--------------+
| id |     address     |     updated_at      |     started_at      | live_bytes | key_bytes | value_bytes | intent_bytes | system_bytes | leader_ranges | repl_ranges | avail_ranges |
+----+-----------------+---------------------+---------------------+------------+-----------+-------------+--------------+--------------+---------------+-------------+--------------+
|  1 | 127.0.0.1:26001 | 2016-01-13 22:06:42 | 2016-01-13 13:06:02 |   12885558 |    144583 |    14436491 |          329 |         2444 |             1 |           1 |            1 |
+----+-----------------+---------------------+---------------------+------------+-----------+-------------+--------------+--------------+---------------+-------------+--------------+
```

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3828)

<!-- Reviewable:end -->

Commit:
Add 'node' command to list nodes and show their status.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=3828) <br/>You should sign our Contributor License Agreement in order to get your pull request merged.<br/>

cuongdo commented:
cc @mrtracy @maxlang

tamird commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cli/http_util.go, line 41 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-4Srg1e7KFKdKg7-p:-K8-4SrhGwjIJLuRhMwM:1701839870) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/http_util.go#L41)):*
this looks identical to some code in acceptance/main_test. could you put it somewhere it could be used by both packages?

---

*[cli/node.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-4tgPwCpYxv2cr0rj:-K8-4tgPwCpYxv2cr0rk:2054919098) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L50)):*
instead of `panicGuard`, should this method just return an error? cc @tschottdorf

---

*[cli/node.go, line 100 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-5IhOLvpdLf75f1dA:-K8-5IhOLvpdLf75f1dB:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L100)):*
ditto

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/http_util.go, line 41 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-4Srg1e7KFKdKg7-p:-K8-LSQtmG_jAqsADY42:1800734390) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/http_util.go#L41)):*
also check if we have something better somewhere. The acceptance tests may just use old code. I thought we already had a method that would create a client for a given context.

---

*[cli/node.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-4tgPwCpYxv2cr0rj:-K8-J6my4RbdUTYt9erS:-259768989) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L50)):*
yes, I wouldn't advise to use `panicGuard` in new code unless it makes a lot of sense (by significantly reducing useless error handling). Actually if you could add a comment to that effect on the def on `panicGuard`, that'd be appreciated.

---

*[cli/node.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-KzmApOFIKsJfiqSE:-K8-KzmApOFIKsJfiqSF:-203240349) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L77)):*
Would be nice to automatically generate those from the proto in `init()` (with light use of reflection). I suppose there are ways to generate an ordering as well (if there's no built-in way, a new struct tag could do it).
Just a thought, may not be worth it.

---

*[cli/node.go, line 88 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-L1iZeCfD3oKZt7dg:-K8-L1i_xfj3idAbzAb-:1304707918) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L88)):*
Maybe we can just abbreviate the original names (in the proto tags) to avoid having to translate here.

---

*[cli/node.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-JhPFweyaaz4pw4Qt:-K8-JhPFweyaaz4pw4Qu:1588617173) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L167)):*
don't feel that this comment and the corresponding one (Node status) are particularly helpful (if you feel strongly about them no problem, but please add in corresponding comments in the other `cli` source files).

---

*[server/status.go, line 629 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-KNeBnS4J2dgS8kEu:-K8-KNeBnS4J2dgS8kEv:-900404838) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/server/status.go#L629)):*
do we already query other endpoints from `cli`? If so, appropriate helpers for all of them should be added along with this one.

---

*[server/status.go, line 634 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-K_ifI71TkeLFg5tG:-K8-K_ifI71TkeLFg5tH:-1834646661) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/server/status.go#L634)):*
since you're passing `nodeID` as a string, I think you can special-case `""` instead of a new method.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[cli/node.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-_b7oeP8we980NxwA:-K8-_b7oeP8we980NxwB:586887070) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L70)):*
We might want a `plain` output format as well so that the output can be used in scripts. Not for this PR, though.

---

*[cli/node.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-KzmApOFIKsJfiqSE:-K8-_GLFxwNGjPb6SrY7:-1484946925) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L77)):*
I'm not sure reflection is worth it here. While reflection could show all the fields in the response, I think we'll want to aggregate some of these fields which will have a much more difficult time fitting into a reflection-based approach.

---

*[cli/node.go, line 151 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-_kHssr5C7RoGGZsP:-K8-_kHssr5C7RoGGZsQ:-1035337605) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L151)):*
This detail is good, though I think users will often want a single count of the bytes on a node. Perhaps the `node status` command should have a `--detail` flag that defaults to false. I also want to know the available space on a node which currently isn't being returned. This is something that can be addressed in a follow-on PR.

---

*[cli/node.go, line 154 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-aaMyS8xds2qVayFo:-K8-aaMyS8xds2qVayFp:1878495830) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L154)):*
Similar to above, I think the default should be simplified to `AvailableRangeCount` and a `--detail` flag should turn on `LeaderRangeCount` and `ReplicatedRangeCount`.

---

*[cli/node.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-JhPFweyaaz4pw4Qt:-K8-a9kpGrt1xMaM7fv4:787511989) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L167)):*
Yeah, this sort of delimiter comment isn't used in the `cli` package.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-:-K8-azFgUBW9MGZXifCo:676834233)*
<!-- Sent from Reviewable.io -->

Commit:
Addressed CircleCI and code review feedback. Also cleaned up some style issues.

cuongdo commented:
Review status: 4 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[cli/node.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-_b7oeP8we980NxwA:-K800_n6D0dGLIEc_FCY:1199982963) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L70)):*
Agreed. MySQL's CLI has some nice output options that make it nice to script and use interactively. We should look into those in the future.

---

*[cli/node.go, line 151 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-_kHssr5C7RoGGZsP:-K801C-q0JfA2enGPufO:1613188253) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L151)):*
Agreed

---

*[cli/node.go, line 154 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-aaMyS8xds2qVayFo:-K801EzkX7IKfGfqYXS-:5173665) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L154)):*
OK

---

*[cli/node.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-JhPFweyaaz4pw4Qt:-K801SK_be7qK8ziytCv:610327451) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L167)):*
Removed

---

*[server/status.go, line 629 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-KNeBnS4J2dgS8kEu:-K8-cciL03xZGZEHZVwP:-1347133946) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/server/status.go#L629)):*
As far as I can tell, we use specialized, non-HTTP clients for the other parts of the CLI (SQL client, KV client, etc.).

---

*[server/status.go, line 634 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-K_ifI71TkeLFg5tG:-K802u0w-vuqk_HidQ_W:-1407218161) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/server/status.go#L634)):*
changed

---

*[cli/http_util.go, line 41 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-4Srg1e7KFKdKg7-p:-K80-yRq1MkedSJHt_MA:-299183068) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/http_util.go#L41)):*
Good points. For the CLI, I'm now using context.GetHTTPClient(). I've moved most of the duplicated getJSON() code to util/http.go. The number of lines I could delete isn't as much as I'd have liked, because acceptance doesn't have a global Context like the CLI does.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 4 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[cli/node.go, line 100 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-5IhOLvpdLf75f1dA:-K806Lox9mAk3RB4wYN6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L100)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[acceptance/cluster/http.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8091XouV_DxCqW11sd:-K8091XouV_DxCqW11se:-469848028) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/acceptance/cluster/http.go#L49)):*
`return util.GetJSON(...)`

---

*[cli/node.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K808fgWcqxoImJw9Izx:-K808fgWcqxoImJw9Izy:-2129630164) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/cli/node.go#L118)):*
i feel like we should exit non-zero here

---

*[cli/start.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K808lQm3tqjb1t1ofYO:-K808lQm3tqjb1t1ofYP:-1438362273) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/cli/start.go#L76)):*
you're already logging this error at the caller; remove this

---

*[cli/start.go, line 80 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K808pfpZCFOx_m8Lqy6:-K808pfqETNDC_GOCl2v:-1858338215) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/cli/start.go#L80)):*
similar to comment above: `return util.GetJSON(...)`, skip the logging

---

*[util/http.go, line 245 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8084GqxO5K7IQcPHXw:-K8084GqxO5K7IQcPHXx:516743574) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/util/http.go#L245)):*
"uses the supplied client to ...."

---

*[util/http.go, line 258 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K80876N3WWa9oxc4c0e:-K80876N3WWa9oxc4c0f:-1594617397) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/util/http.go#L258)):*
`return json.Unmarshal(b, v)`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

Commit:
Further code review fixes

cuongdo commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[acceptance/cluster/http.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8091XouV_DxCqW11sd:-K80Smg3tRuDCkr0Acq9:-1117124323) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/acceptance/cluster/http.go#L49)):*
The main intention was to log an error at verbosity 1, which returning doesn't do. I saw it done in the old code, but I'm not sure if that's still a current convention.

---

*[cli/node.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8-JhPFweyaaz4pw4Qt:-K80TUfwgbe-R3cNwRpU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/df3a1866a9cc84addb6ca2ad7b1cba7000a697a0/cli/node.go#L167)):*
Done.

---

*[cli/node.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K808fgWcqxoImJw9Izx:-K80TSWlt1Olhkk7RZTQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/cli/node.go#L118)):*
Done.

---

*[cli/start.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K808lQm3tqjb1t1ofYO:-K80UMGR60wLcbqaHFmR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/cli/start.go#L76)):*
Done.

---

*[cli/start.go, line 80 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K808pfpZCFOx_m8Lqy6:-K80TuxXBqQftXYRwtOT:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/cli/start.go#L80)):*
Done.

---

*[util/http.go, line 245 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8084GqxO5K7IQcPHXw:-K80U4pAXLal6JxPGDE9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/util/http.go#L245)):*
Done.

---

*[util/http.go, line 258 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K80876N3WWa9oxc4c0e:-K80UFYP6SWAKCxk18y6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/util/http.go#L258)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[acceptance/cluster/http.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8091XouV_DxCqW11sd:-K80WvJPWwYYEm2bjAY2:229258881) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/acceptance/cluster/http.go#L49)):*
If this is important to log, the caller should do it. currently this code is swallowing this error, which seems Bad™

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

Commit:
More code review issues

cuongdo commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[acceptance/cluster/http.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8091XouV_DxCqW11sd:-K80lVDur7JYtP862-m8:-1689220901) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/acceptance/cluster/http.go#L49)):*
Ah yes, that was unintentional.

done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 8 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[acceptance/cluster/http.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8091XouV_DxCqW11sd:-K80mfgOCIv5E3HYSgYl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/acceptance/cluster/http.go#L49)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Is there a test for this command? It seems like something that could break without us noticing, for example if the node endpoint or the data format returned from the server are changed.

---

Review status: 8 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[acceptance/cluster/http.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8091XouV_DxCqW11sd:-K80knQJwyZa-zKF8b3L:810981944) ([raw file](https://github.com/cockroachdb/cockroach/blob/7451757ab4c6d4151313e262bc18b347b2784ba8/acceptance/cluster/http.go#L49)):*
It actually looks like an oversight in the previous code. An http error would correctly return an error, but a json marshalling error was being swallowed. There's almost no way that was intentional.

I don't necessarily agree that logging should be at a higher level; in some cases the source of an error is ambiguous at the highest level, and logging closer to the source of an error can often provide clues to the underlying cause.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-:-K80oEwnd8VKbZGJOOxA:-446741742)*
<!-- Sent from Reviewable.io -->

tamird commented:
Agreed that a test is a good idea.

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-:-K80ok43KNQ66qEGHhNW:2100279216)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Not yet. I'm not sure how to approach testing for this command. The test server created by the CLI tests doesn't have any nodes. So, I could add nodes, but I'm unclear how deep this rabbit hole is. Maybe add an acceptance test? Suggestions appreciated.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-:-K806YzUibJniSPc3-yM:-1057845609)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Why does the test server not have any nodes? I would think you could model this on the `Example_*` style tests. See `Example_basic` in `cli/cli_test.go`.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-:-K80rfoKXZ8tly70wx3c:-208102566)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
I've already discussed this with Cuong, but for posterity: the server *does* have a node, but the "nodes" information is based on routinely recorded status summaries, and the first status summary is not immediately recorded. In order to test this, you need to force the server to synchronously record status summaries; there are already methods available for this, so at most an extra method would need to be added to TestServer.  

For an example of forcing summaries to be written, see TestStatusSummaries in server/node_test.go.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-:-K8164PNQj7mJYeyOf8U:1680424835)*
<!-- Sent from Reviewable.io -->

Commit:
Added tests for CLI.

cuongdo commented:
I've added some tests for the new CLI functionality. Also, I've added `RunWithCapture`, which captures the output of an execution of the CLI and should be useful for other CLI tests.

---

Review status: 7 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-:-K857EUDkKbLcbT5Oxt-:316908764)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 64 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85ALRQ-zY3w9jWyiiz:-K85ALRQ-zY3w9jWyij-:1797707716) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L64)):*
replace "the CLI" with "c" or "the receiver"

---

*[cli/cli_test.go, line 90 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85B7NECfgOI8Cm_l1s:-K85B7NECfgOI8Cm_l1t:1247166722) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L90)):*
this is odd. can't you unconditionally `outC <- captureResult{buf.String(), err}`?

---

*[cli/cli_test.go, line 104 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85BTJft_RTBLSx8Wxk:-K85BTJft_RTBLSx8Wxl:-1122233435) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L104)):*
this is also kind of odd. why not the simpler:
```go
out, err = outResult.out, outResult.err
if x := recover(); x != nil {
  err = util.Errorf("panic: %v", x)
}
```

---

*[cli/cli_test.go, line 112 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85BguFfCey1q-YGspo:-K85BguFfCey1q-YGspp:833770320) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L112)):*
replace the "the CLI" here. running the command line interface doesn't sounds odd

---

*[cli/cli_test.go, line 608 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85A0-7XHxbcYq8x9bo:-K85A0-7XHxbcYq8x9bp:1079461155) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L608)):*
looks like the convention is to have an empty line between the code and the `Output` section. Also, this is going to have to be rebased on top of #3859 if that one gets merged first.

---

*[cli/cli_test.go, line 632 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85Bw4c_3OWmZW-H4WY:-K85Bw4c_3OWmZW-H4WZ:308099403) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L632)):*
you probably want `t.Fatal` here; if the command failed, there's no sense in continuing the test.

---

*[cli/cli_test.go, line 638 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85C06MyGKomG8jM_f6:-K85C06MyGKomG8jM_f7:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L638)):*
ditto

---

*[cli/cli_test.go, line 651 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85CdQ63PuxZuPu_-Q-:-K85CdQ63PuxZuPu_-Q0:65719104) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L651)):*
you might get a better error message if you include `s.Err()` here.

---

*[cli/cli_test.go, line 658 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85ChkHiMNlH7DjaOry:-K85ChkHiMNlH7DjaOrz:2085412223) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L658)):*
you can't assume this is EOF; you need to check `s.Err()`

---

*[cli/cli_test.go, line 669 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85ClZxo8giBUGslJWk:-K85C_d2iqWe_JFC-qGw:1352623403) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L669)):*
you can't assume this is EOF; you need to check `s.Err()`ditto

---

*[cli/cli_test.go, line 672 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85Coqpy_4xp0n9xcwi:-K85Coqpy_4xp0n9xcwj:263472121) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L672)):*
tip:
```
if a, e := len(fields), len(nodeColumnHeaders); a != e {
  t.Fatalf("...", a, e)
}
```

---

*[cli/cli_test.go, line 687 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85D5I5zjkRq3rTaZJg:-K85D5I5zjkRq3rTaZJh:634538267) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L687)):*
nit: put fields[1]` on the right for consistency with the above. also, consider using the `if a,e := ...; a != e { .. }` pattern throughout

---

*[cli/cli_test.go, line 730 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85DMwb_mAXQAmdkmYK:-K85DMwb_mAXQAmdkmYL:329208845) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L730)):*
use s.Err()

---

*[cli/cli_test.go, line 732 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85DZyjPY-Esfdud9un:-K85DZyjPY-Esfdud9uo:1656826363) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L732)):*
consider extracting this regexp into a package-local variable using `regexp.MustCompile`; that'll save you from having to deal with the `err` return here.

---

*[cli/cli_test.go, line 746 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85Dyrsg1aqA4h1F1v9:-K85Dyrsg1aqA4h1F1vA:1971939085) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L746)):*
https://godoc.org/time#Time.Truncate ?

---

*[cli/cli_test.go, line 748 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EW00aj4u2UuxdkIv:-K85EW00aj4u2UuxdkIw:-1107227933) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L748)):*
nit: typically format errors with `%s`

---

*[cli/cli_test.go, line 757 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85E2qwVnIODD4CARn2:-K85E2qwVnIODD4CARn3:2092130113) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L757)):*
i think this would be clearer if `secs` was a `time.Duration`. you may want to call it `delta`, for clarity.

---

*[cli/cli_test.go, line 759 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EMKCzbzi-vMR0ZAZ:-K85EMKCzbzi-vMR0ZA_:497739299) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L759)):*
I think this range is actually `(%v, %v)`. Also, `time.Time` implements stringer, so you can use `%s` to format these values.

---

*[cli/cli_test.go, line 769 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EcyqazPAbPeN83F5:-K85EcyqazPAbPeN83F6:-915558125) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L769)):*
What does this throw away? Also, `https://godoc.org/strings#Fields` might make sense here (after replacing all `'|'` with whitespace characters).

---

*[cli/node.go, line 117 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K859Hsz5NOiFSmflXzG:-K859Ht-lyBTlpggpFqu:2025395221) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/node.go#L117)):*
add "TODO(name):" here and please file an issue after this is merged

---

*[server/testserver.go, line 295 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K859cQ6sfxY2luhEDB5:-K859cQ6sfxY2luhEDB6:-2024248228) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/server/testserver.go#L295)):*
might be more explicit as `ts.Server.writeSummaries()`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 669 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85ClZxo8giBUGslJWk:-K85F1Qlp19BpbZFJUEh:-2107085695) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L669)):*
err, this should have just been "ditto"

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

Commit:
Commit WIP

Commit:
More code review changes

cuongdo commented:
Review status: 6 of 10 files reviewed at latest revision, 25 unresolved discussions.

---

*[cli/cli_test.go, line 64 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85ALRQ-zY3w9jWyiiz:-K85b7kErUmLvZY1HTV4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L64)):*
Done.

---

*[cli/cli_test.go, line 90 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85B7NECfgOI8Cm_l1s:-K85bcVN_KW9zjAef48d:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L90)):*
Done.

---

*[cli/cli_test.go, line 104 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85BTJft_RTBLSx8Wxk:-K85c3oYRf2n7ehOF5Zp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L104)):*
Done.

---

*[cli/cli_test.go, line 112 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85BguFfCey1q-YGspo:-K85cL1EiuURBLQj8qtU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L112)):*
Done.

---

*[cli/cli_test.go, line 608 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85A0-7XHxbcYq8x9bo:-K85eSZWRAQn277sup-_:1681699936) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L608)):*
Thanks for the heads up. Rebased.

---

*[cli/cli_test.go, line 632 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85Bw4c_3OWmZW-H4WY:-K85eaFu02bXFFjCNhyC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L632)):*
Done.

---

*[cli/cli_test.go, line 638 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85C06MyGKomG8jM_f6:-K85eb00g3fprdz64by0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L638)):*
Done.

---

*[cli/cli_test.go, line 651 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85CdQ63PuxZuPu_-Q-:-K85f1byt0gOLrIJyEeI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L651)):*
Done.

---

*[cli/cli_test.go, line 658 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85ChkHiMNlH7DjaOry:-K85f21MvXD9osxerSoq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L658)):*
Done.

---

*[cli/cli_test.go, line 669 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85ClZxo8giBUGslJWk:-K85f9JWhT17BjAFAJOX:2129385488) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L669)):*
al,Done.

---

*[cli/cli_test.go, line 672 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85Coqpy_4xp0n9xcwi:-K85g28g1tMHHTnTrurU:766161033) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L672)):*
Nice tip! Thanks.

---

*[cli/cli_test.go, line 687 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85D5I5zjkRq3rTaZJg:-K85giD_iJmnxB2VQIze:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L687)):*
Done.

---

*[cli/cli_test.go, line 730 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85DMwb_mAXQAmdkmYK:-K85gqtgQnWLgSZsZvrO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L730)):*
Done.

---

*[cli/cli_test.go, line 732 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85DZyjPY-Esfdud9un:-K85hU6YSQRDAW_csycQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L732)):*
Done.

---

*[cli/cli_test.go, line 746 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85Dyrsg1aqA4h1F1v9:-K85icNlXMCByFwDXrJp:720839233) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L746)):*
The problem is that the start time has a local time zone, while the time read from the CLI output is in UTC (because the time zone is unknown). One of the side effects of the call to time.Parse(..., start.Format(...)) is that the time zone information is lost and is therefore assumed to be UTC.

---

*[cli/cli_test.go, line 748 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EW00aj4u2UuxdkIv:-K85jGM9nu60hSjDR3GD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L748)):*
Done.

---

*[cli/cli_test.go, line 757 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85E2qwVnIODD4CARn2:-K85jkO7A_KZGHZOZE3o:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L757)):*
Done.

---

*[cli/cli_test.go, line 759 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EMKCzbzi-vMR0ZAZ:-K85jzi3fnyIMQx7ApUB:567269527) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L759)):*
Fixed format string. I think the range is [%s,%s], because neither Before nor After are inclusive. So, if tm == start, tm.Before(start) would fail. Therefore, start is part of the range. If tm == end, similar logic applies.

---

*[cli/cli_test.go, line 769 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EcyqazPAbPeN83F5:-K85lc0d1yWaseLFXHMo:1342690521) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L769)):*
The output looks like this:

| 1 | 2 | this is one column |

Fields won't work, because it would return "this", "is", "one", "column" all as separate fields. This splits based on '|' and throws away surrounding whitespace. Not 100% correct in all cases, but for CLI tests, we don't expect surrounding spaces anyway.

---

*[cli/node.go, line 117 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K859Hsz5NOiFSmflXzG:-K85ZnHrfKmFffyjLwC0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/node.go#L117)):*
Done.

---

*[server/testserver.go, line 295 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K859cQ6sfxY2luhEDB5:-K85_3ySFVQubMCPvQJV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/server/testserver.go#L295)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[cli/cli_test.go, line 746 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85Dyrsg1aqA4h1F1v9:-K8660OZOQjuBQKuJ9Fi:-553481103) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L746)):*
Could you solve this by replacing `time.Parse(localTimeFormat, timeStr)` with `time.ParseInLocation(localTimeFormat, timeStr, start.Location())`?

---

*[cli/cli_test.go, line 748 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EW00aj4u2UuxdkIv:-K865OmK32O03iLubUqJ:1449476315) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L748)):*
Looks like this one got missed.

---

*[cli/cli_test.go, line 759 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EMKCzbzi-vMR0ZAZ:-K865TnEb6CcF32nGpxo:-1605280730) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L759)):*
Yep, you're right, my mistake.

---

*[cli/cli_test.go, line 769 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EcyqazPAbPeN83F5:-K865ZSF8vwAg7zd6YFE:-2065850108) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L769)):*
OK, I understand `strings.Fields` won't work, but why is it expected that you'll see empty strings here, and that it is OK to discard them? Also, rather than trimming spaces, why not split on " | "?

---

*[cli/cli_test.go, line 675 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8651UJbAiwl7qbY5dY:-K8651UJbAiwl7qbY5dZ:-1974879554) ([raw file](https://github.com/cockroachdb/cockroach/blob/84a9fda78ddbd4a1e3bf890a07507725192166ad/cli/cli_test.go#L675)):*
you don't need to `time.Duration()` the `5`; constants can be untyped in Go, so `5*time.Second` also works and reads nicer.

---

*[cli/cli_test.go, line 717 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K865AQXcvZUJFPZCQpn:-K865AQXcvZUJFPZCQpo:1777759279) ([raw file](https://github.com/cockroachdb/cockroach/blob/84a9fda78ddbd4a1e3bf890a07507725192166ad/cli/cli_test.go#L717)):*
tip: `.Text()` allocates a string. it doesn't really matter here, but generally you'd want to do that just once with `if foo := s.Text(); ... { ... }`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

Commit:
more code review fixes

cuongdo commented:
Review status: 9 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[cli/cli_test.go, line 746 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85Dyrsg1aqA4h1F1v9:-K86F6AVnq0W9rwOBYmU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L746)):*
Done.

---

*[cli/cli_test.go, line 769 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EcyqazPAbPeN83F5:-K86FdaUDx6ioqav383e:949396490) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L769)):*
Splitting on `" | "` won't work, because the `"|"` in the first column has no space to the left of it, and the final `"|"` has no `" "` to the right of it. So, splitting on `" | "` will return `[]string{"| 1", "2", "this is one column |"}`.

Example:
https://play.golang.org/p/D9geZHJMeQ

---

*[cli/cli_test.go, line 675 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K8651UJbAiwl7qbY5dY:-K86DzAzFP4AtG617STG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/84a9fda78ddbd4a1e3bf890a07507725192166ad/cli/cli_test.go#L675)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[cli/cli_test.go, line 769 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EcyqazPAbPeN83F5:-K86w93-T4QcSepz53tL:-660064535) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L769)):*
I see. Still:

> why is it expected that you'll see empty strings here, and why it is OK to discard them?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

Commit:
more code review fixes

cuongdo commented:
Review status: 9 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[cli/cli_test.go, line 769 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EcyqazPAbPeN83F5:-K87ZAZ0oecVaI0VwVCw:-271806825) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L769)):*
Split() generates empty strings for the 1st and last entries, because the delimiter is the first and last character of its input. I've made the code handle this more cleanly.

To be clear, here's a simple demo of the Split() behavior:
https://play.golang.org/p/6BtnTOkCGY

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM

---

Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 769 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-K85EcyqazPAbPeN83F5:-K886Ol1udaewR9YKXx4:1642398675) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01ec49dda0dc56a6f8ee392272c2dab81226fed/cli/cli_test.go#L769)):*
Cool, thanks for clarifying that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3828#-:-K886nyO8Mdf6Vr_Xfvc:676834233)*
<!-- Sent from Reviewable.io -->
